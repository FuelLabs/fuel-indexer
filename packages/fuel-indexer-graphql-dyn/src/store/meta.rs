use super::self_prelude::*;
use convert_case::{Case, Casing, StateConverter};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Name {
    Pascal(String),
    Snake(String),
    Camel(String),
}

impl Name {
    pub fn new_pascal(name: impl Into<String>) -> Self {
        Self::Pascal(name.into())
    }
    pub fn new_snake(name: impl Into<String>) -> Self {
        Self::Snake(name.into())
    }
    pub fn new_camel(name: impl Into<String>) -> Self {
        Self::Camel(name.into())
    }

    pub fn to_converter(&self) -> StateConverter<String> {
        match self {
            Name::Pascal(name) => name.from_case(Case::Pascal),
            Name::Snake(name) => name.from_case(Case::Snake),
            Name::Camel(name) => name.from_case(Case::Camel),
        }
    }
    pub fn to_pascal_string(&self) -> String {
        self.to_converter().to_case(Case::Pascal)
    }
    pub fn to_snake_string(&self) -> String {
        self.to_converter().to_case(Case::Snake)
    }
    pub fn to_camel_string(&self) -> String {
        self.to_converter().to_case(Case::Camel)
    }
}

impl fmt::Display for Name {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Name::Pascal(name) => write!(f, "{}", name),
            Name::Snake(name) => write!(f, "{}", name),
            Name::Camel(name) => write!(f, "{}", name),
        }
    }
}
