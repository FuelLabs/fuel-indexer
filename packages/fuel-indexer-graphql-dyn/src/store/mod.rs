pub mod assoc;
pub mod assoc_type;
pub mod data;
pub mod data_type;
pub mod meta;
pub mod obj;
pub mod obj_type;
#[allow(clippy::module_inception)]
pub mod store;
pub mod store_type;
pub mod store_type_builder;

pub(self) mod self_prelude {
    pub use super::super::self_prelude::*;
    pub use indexmap::IndexMap;
    pub use serde::{Deserialize, Serialize};
    pub use std::fmt;
    pub use std::str::FromStr;
}

pub use assoc::*;
pub use assoc_type::*;
pub use data::*;
pub use data_type::*;
pub use meta::*;
pub use obj::*;
pub use obj_type::*;
pub use store::*;
pub use store_type::*;
pub use store_type_builder::*;
