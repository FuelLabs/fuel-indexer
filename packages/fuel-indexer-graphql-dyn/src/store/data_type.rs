use super::meta::*;
use super::self_prelude::*;

#[derive(Clone, Debug, Serialize, Deserialize, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub enum DataTypeId {
    Unit,
    Bool,
    U8,
    U16,
    U32,
    U64,
    B256,
    Byte,
    Bytes,
    String,
    Array(Box<DataTypeId>),
    Name(Name),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum DataType {
    Unit,
    Bool,
    U8,
    U16,
    U32,
    U64,
    B256,
    Byte,
    Bytes,
    String,
    Array(Box<DataTypeId>),
    Composite(Name, IndexMap<DataFieldId, DataFieldType>),
    Enum(Name, IndexMap<DataFieldId, DataFieldType>),
}

pub type DataFieldId = String;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DataFieldType {
    pub name: Name,
    pub type_id: DataTypeId,
}

impl DataType {
    pub fn name(&self) -> Name {
        match self {
            Self::Unit => Name::new_pascal("Unit"),
            Self::Bool => Name::new_pascal("Boolean"),
            Self::U8 => Name::new_pascal("U8"),
            Self::U16 => Name::new_pascal("U16"),
            Self::U32 => Name::new_pascal("U32"),
            Self::U64 => Name::new_pascal("U64"),
            Self::B256 => Name::new_pascal("B256"),
            Self::Byte => Name::new_pascal("Byte"),
            Self::Bytes => Name::new_pascal("Bytes"),
            Self::String => Name::new_pascal("String"),
            Self::Array(ty) => Name::new_pascal(format!("[{}]", ty)),
            Self::Composite(name, _) => name.clone(),
            Self::Enum(name, _) => name.clone(),
        }
    }
}

impl FromStr for DataTypeId {
    type Err = anyhow::Error;
    fn from_str(str: &str) -> anyhow::Result<Self, Self::Err> {
        match str {
            "Unit" => Ok(Self::Unit),
            "Boolean" => Ok(Self::Bool),
            "U8" => Ok(Self::U8),
            "U16" => Ok(Self::U16),
            "U32" => Ok(Self::U32),
            "U64" => Ok(Self::U64),
            "B256" => Ok(Self::B256),
            "Byte" => Ok(Self::Byte),
            "Bytes" => Ok(Self::Bytes),
            "String" => Ok(Self::String),
            str if str.starts_with("Array__") => {
                todo!()
            }
            str => {
                let name = str.to_string();
                Ok(Self::Name(Name::new_pascal(name)))
            }
        }
    }
}

impl std::fmt::Display for DataTypeId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Unit => write!(f, "Unit"),
            Self::Bool => write!(f, "Boolean"),
            Self::U8 => write!(f, "U8"),
            Self::U16 => write!(f, "U16"),
            Self::U32 => write!(f, "U32"),
            Self::U64 => write!(f, "U64"),
            Self::B256 => write!(f, "B256"),
            Self::Byte => write!(f, "Byte"),
            Self::Bytes => write!(f, "Bytes"),
            Self::String => write!(f, "String"),
            Self::Array(ty) => write!(f, "[{}]", ty),
            Self::Name(id) => write!(f, "{}", id),
        }
    }
}
