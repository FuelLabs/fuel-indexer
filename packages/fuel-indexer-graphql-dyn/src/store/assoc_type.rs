use super::data_type::*;
use super::meta::*;
use super::obj_type::*;
use super::self_prelude::*;

pub type AssocTypeId = String;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AssocType {
    pub name: Name,
    pub tail: ObjTypeId,
    pub head: ObjTypeId,
    pub fields: IndexMap<DataFieldId, DataFieldType>,
}
