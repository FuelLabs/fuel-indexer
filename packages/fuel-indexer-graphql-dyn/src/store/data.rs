use super::data_type::*;
use super::self_prelude::*;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum Data {
    Json(serde_json::Value),
}

impl Data {
    pub fn r#type(&self) -> &DataTypeId {
        todo!()
    }

    pub fn value(&self) -> &serde_json::Value {
        match self {
            Self::Json(json) => json,
        }
    }
}

impl From<serde_json::Value> for Data {
    fn from(value: serde_json::Value) -> Self {
        Self::Json(value)
    }
}
