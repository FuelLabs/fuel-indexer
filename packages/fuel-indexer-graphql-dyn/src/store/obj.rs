use super::data::*;
use super::obj_type::*;
use super::self_prelude::*;

pub type ObjId = String;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Obj(pub ObjTypeId, pub Data);

impl Obj {
    pub fn new(obj_type_id: ObjTypeId, data: Data) -> Self {
        Self(obj_type_id, data)
    }

    pub fn type_id(&self) -> &ObjTypeId {
        &self.0
    }
    pub fn data(&self) -> &Data {
        &self.1
    }
}
