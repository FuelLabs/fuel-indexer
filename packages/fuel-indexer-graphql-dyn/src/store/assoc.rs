use super::assoc_type::*;
use super::data::*;
use super::obj::*;
use super::self_prelude::*;

pub type AssocTime = u64;

#[derive(Clone, Debug, Serialize, Deserialize, Hash, PartialEq, Eq)]
pub struct AssocKey(pub ObjId, pub AssocTypeId);

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Assoc(pub ObjId, pub AssocTime, pub Data);

impl Assoc {
    pub fn id(&self) -> &ObjId {
        &self.0
    }
    pub fn time(&self) -> &AssocTime {
        &self.1
    }
    pub fn data(&self) -> &Data {
        &self.2
    }
}
