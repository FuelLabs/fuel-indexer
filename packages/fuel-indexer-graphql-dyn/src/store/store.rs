use super::assoc::*;
use super::obj::*;
use super::self_prelude::*;
use super::store_type::*;

type Result<T> = anyhow::Result<T, StoreError>;

/// See: https://research.facebook.com/publications/tao-facebooks-distributed-data-store-for-the-social-graph/
#[async_trait]
pub trait Store: Send + Sync {
    /// Returns the type of the store.
    fn r#type(&self) -> &StoreType;
    //
    /// Gets a single object by its ID.
    async fn obj_get(&self, id: &ObjId) -> Result<Option<Obj>>;
    /// Gets objects by their IDs.
    async fn obj_get_many(&self, ids: &[ObjId]) -> Result<Vec<Option<Obj>>>;
    // Gets associations of an object by their head IDs.
    async fn assoc_get(
        &self,
        key: &AssocKey,
        ids: &[ObjId],
        high: Option<AssocTime>,
        low: Option<AssocTime>,
    ) -> Result<Vec<Option<Assoc>>>;
    // Gets the count of associations an object has.
    async fn assoc_count(&self, key: &AssocKey) -> Result<u64>;
    // Gets associations of an object.
    async fn assoc_range(
        &self,
        key: &AssocKey,
        offset: u64,
        limit: u64,
    ) -> Result<Vec<Assoc>>;
    // Gets associations of an object in the given time range.
    async fn assoc_time_range(
        &self,
        key: &AssocKey,
        high: AssocTime,
        low: AssocTime,
        limit: u64,
    ) -> Result<Vec<Assoc>>;
}

pub type StoreResult<T> = anyhow::Result<T, StoreError>;

#[derive(thiserror::Error, Debug)]
pub enum StoreError {
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}
