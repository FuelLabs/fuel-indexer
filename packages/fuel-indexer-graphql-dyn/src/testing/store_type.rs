use crate::store::*;

pub fn new_test_store_type() -> Result<StoreType, ()> {
    let mut store_type = StoreTypeBuilder::default();

    let (call_receipt, data) = store_type.define_composite("CallReceiptData");
    data.define_field("param1", &DataTypeId::U64);
    data.define_field("param2", &DataTypeId::U64);

    let (return_receipt, data) = store_type.define_composite("ReturnReceiptData");
    data.define_field("val", &DataTypeId::U64);

    let (receipt_type, data) = store_type.define_enum("ReceiptType");
    data.define_variant("CallReceipt", &DataTypeId::Unit);
    data.define_variant("ReturnReceipt", &DataTypeId::Unit);

    let (receipt_data, data) = store_type.define_enum("ReceiptData");
    data.define_variant("CallReceiptData", &call_receipt);
    data.define_variant("ReturnReceiptData", &return_receipt);

    let (chain, obj) = store_type.define_obj("Chain");
    obj.define_field("number", &DataTypeId::U64);

    let (block, obj) = store_type.define_obj("Block");
    obj.define_field("number", &DataTypeId::U64);
    obj.define_field("hash", &DataTypeId::B256);
    obj.define_field("parent_hash", &DataTypeId::B256);

    let (transaction, obj) = store_type.define_obj("Transaction");
    obj.define_field("index", &DataTypeId::U64);
    obj.define_field("hash", &DataTypeId::B256);
    obj.define_field("gas_limit", &DataTypeId::U64);
    obj.define_field("gas_price", &DataTypeId::U64);
    obj.define_field("block_hash", &DataTypeId::B256);

    let (_receipt, obj) = store_type.define_obj("Receipt");
    obj.define_field("type", &receipt_type);
    obj.define_field("data", &receipt_data);

    let (_chain_has_block, _assoc) = store_type.define_assoc(&chain, "Has", &block);
    let (_chain_has_transaction, _assoc) =
        store_type.define_assoc(&chain, "Has", &transaction);
    let (_block_has_transaction, _assoc) =
        store_type.define_assoc(&block, "Has", &transaction);

    store_type.finish()
}

#[test]
fn test_store_type() {
    use insta::*;

    let store_type = new_test_store_type().unwrap();
    assert_ron_snapshot!(store_type);
}
