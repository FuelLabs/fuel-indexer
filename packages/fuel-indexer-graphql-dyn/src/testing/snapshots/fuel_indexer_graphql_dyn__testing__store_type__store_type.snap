---
source: packages/fuel-indexer-graphql-dyn/src/testing/store_type.rs
expression: store_type
---
StoreType(
  data: {
    Unit: Unit,
    Bool: Bool,
    U8: U8,
    U16: U16,
    U32: U32,
    U64: U64,
    B256: B256,
    Byte: Byte,
    Bytes: Bytes,
    String: String,
    Name(Pascal("CallReceiptData")): Composite(Pascal("CallReceiptData"), {
      "param1": DataFieldType(
        name: Snake("param1"),
        type_id: U64,
      ),
      "param2": DataFieldType(
        name: Snake("param2"),
        type_id: U64,
      ),
    }),
    Name(Pascal("ReturnReceiptData")): Composite(Pascal("ReturnReceiptData"), {
      "val": DataFieldType(
        name: Snake("val"),
        type_id: U64,
      ),
    }),
    Name(Pascal("ReceiptType")): Enum(Pascal("ReceiptType"), {
      "call_receipt": DataFieldType(
        name: Pascal("CallReceipt"),
        type_id: Unit,
      ),
      "return_receipt": DataFieldType(
        name: Pascal("ReturnReceipt"),
        type_id: Unit,
      ),
    }),
    Name(Pascal("ReceiptData")): Enum(Pascal("ReceiptData"), {
      "call_receipt_data": DataFieldType(
        name: Pascal("CallReceiptData"),
        type_id: Name(Pascal("CallReceiptData")),
      ),
      "return_receipt_data": DataFieldType(
        name: Pascal("ReturnReceiptData"),
        type_id: Name(Pascal("ReturnReceiptData")),
      ),
    }),
  },
  obj: {
    "Chain": ObjType(
      name: Pascal("Chain"),
      fields: {
        "number": DataFieldType(
          name: Snake("number"),
          type_id: U64,
        ),
      },
    ),
    "Block": ObjType(
      name: Pascal("Block"),
      fields: {
        "number": DataFieldType(
          name: Snake("number"),
          type_id: U64,
        ),
        "hash": DataFieldType(
          name: Snake("hash"),
          type_id: B256,
        ),
        "parent_hash": DataFieldType(
          name: Snake("parent_hash"),
          type_id: B256,
        ),
      },
    ),
    "Transaction": ObjType(
      name: Pascal("Transaction"),
      fields: {
        "index": DataFieldType(
          name: Snake("index"),
          type_id: U64,
        ),
        "hash": DataFieldType(
          name: Snake("hash"),
          type_id: B256,
        ),
        "gas_limit": DataFieldType(
          name: Snake("gas_limit"),
          type_id: U64,
        ),
        "gas_price": DataFieldType(
          name: Snake("gas_price"),
          type_id: U64,
        ),
        "block_hash": DataFieldType(
          name: Snake("block_hash"),
          type_id: B256,
        ),
      },
    ),
    "Receipt": ObjType(
      name: Pascal("Receipt"),
      fields: {
        "type": DataFieldType(
          name: Snake("type"),
          type_id: Name(Pascal("ReceiptType")),
        ),
        "data": DataFieldType(
          name: Snake("data"),
          type_id: Name(Pascal("ReceiptData")),
        ),
      },
    ),
  },
  assoc: {
    "ChainHasBlock": AssocType(
      name: Pascal("ChainHasBlock"),
      tail: "Chain",
      head: "Block",
      fields: {},
    ),
    "ChainHasTransaction": AssocType(
      name: Pascal("ChainHasTransaction"),
      tail: "Chain",
      head: "Transaction",
      fields: {},
    ),
    "BlockHasTransaction": AssocType(
      name: Pascal("BlockHasTransaction"),
      tail: "Block",
      head: "Transaction",
      fields: {},
    ),
  },
)
